### PR TITLE
Eigen3.3 compatibility

### PR DESCRIPTION
**Pull request by ithron**
Fixed some errors that appear when using Eigen3.3.
Still compatible with Eigen 3.2

### CODE REVIEW

**Comment by ithron** [comment]
Note that this PR does not touch issue https://github.com/libigl/libigl/pull/801 (DynamicSparseMatrix deprecation). Only *hard* errors were fixed.

**Comment by jdumas** [comment]
Does `SparseMatrix::StorageIndex` already exists in Eigen 3.2? Shouldn't code use `SparseMatrix::Index` with Eigen <= 3.2?

**Comment by ithron** [comment]
It exists but is undocumented.

**Comment by jdumas** [comment]
Do you know which in version it was introduced? Otherwise, we can go with a `#if EIGEN_VERSION_AT_LEAST(3,3,0)` to conditionally compile the right template instantiation.

File: include/igl/sparse_cached.cpp [hunk 1]
```diff
@@ -117,6 +117,12 @@ IGL_INLINE void igl::sparse_cached(
 
 
 #ifdef IGL_STATIC_LIBRARY
-template void igl::sparse_cached<double>(std::vector<Eigen::Triplet<double, Eigen::SparseMatrix<double, 0, int>::StorageIndex>, std::allocator<Eigen::Triplet<double, Eigen::SparseMatrix<double, 0, int>::StorageIndex> > > const&, Eigen::Matrix<int, -1, 1, 0, -1, 1> const&, Eigen::SparseMatrix<double, 0, int>&);
-template void igl::sparse_cached_precompute<double>(std::vector<Eigen::Triplet<double, Eigen::SparseMatrix<double, 0, int>::StorageIndex>, std::allocator<Eigen::Triplet<double, Eigen::SparseMatrix<double, 0, int>::StorageIndex> > > const&, Eigen::Matrix<int, -1, 1, 0, -1, 1>&, Eigen::SparseMatrix<double, 0, int>&);
+#if EIGEN_VERSION_AT_LEAST(3,2,9)
```
Comment (jdumas) [inline include/igl/sparse_cached.cpp]: May be better to start at v3.3, where the type has been documented.

**Comment by ithron** [comment]
It's documented since [3.3.2](http://eigen.tuxfamily.org/index.php?title=ChangeLog#Eigen_3.3.2)

**Comment by jdumas** [comment]
I think it's fine to start at v3.3.0 actually. See [Eigen's changelog](http://eigen.tuxfamily.org/index.php?title=ChangeLog#Eigen_3.2.9) and note about [Index type here](http://eigen.tuxfamily.org/index.php?title=3.3#Index_typedef). The thing is that `SparseMatrix::Index` is defined as an `int` in the 3.2 branch, but not in the 3.3 branch. So code that is using Eigen 3.3.0 or higher should use `StorageIndex` instead. Sorry for the confusion and thanks for your patience :p

File: include/igl/vertex_triangle_adjacency.cpp [hunk 2]
```diff
@@ -70,7 +70,7 @@ IGL_INLINE void igl::vertex_triangle_adjacency(
   // vfd now acts as a counter
   vfd = NI;
 
-  VF = Eigen::VectorXi(3*F.rows());
+  static_cast<DerivedVF &>(VF )= Eigen::VectorXi(3*F.rows());
```
Comment (jdumas) [inline include/igl/vertex_triangle_adjacency.cpp]: I'm sorry can you remind me why is this line necessary? If you need to access the derived type you can write `VF.derived() = `. I remember having issues with templated types and allocation in Eigen 3.3, not sure what's the typical way to go.

**Comment by ithron** [comment]
It's required because operator= of PlainObjectBase is protected in 3.3. However, `VF.derived() = ...`
does the same. Maybe using `VF.derived()` is better safer.

**Comment by jdumas** [comment]
Right. Maybe we should be using `MatrixBase` instead of `PlainObjectBase` instead then? See [this note](http://libigl.github.io/libigl/style-guidelines/#templating-with-eigen) on the libigl website. @alecjacobson I never remember which one should be used as template, and I don't think we have a FAQ entry on that. I think it got discussed in previous issues but I don't remember which ones. Can you refresh my memory on that?

**Comment by ithron** [comment]
Related: https://github.com/libigl/libigl/pull/414

**Comment by ithron** [comment]
From ["The class hierarchy](https://eigen.tuxfamily.org/dox/TopicClassHierarchy.html):

> These classes serve as base classes for the five core classes mentioned above. They are more internal and so less interesting for users of the Eigen library.

- PlainObjectBase
- DenseCoeffsBase
- EigenBase

So I think either MatrixBase (matrix expression), VectorBase (vector expression) or DenseBase (matrix or vector expression) should be preferred over PlainObjectBase.

But maybe that's beyond the scope of this PR?

**Comment by jdumas** [comment]
@danielepanozzo Can we merge this?
